### PR TITLE
chore: passing consistency token using request properties in HARouting

### PR DESCRIPTION
**Pull request by patrickstuedi**
Following the practice introduced in #8264 this PR is using request properties to pass the consistency token between servers

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java [hunk 1]
```diff
@@ -336,10 +336,14 @@ private static void forwardTo(
         .map(location -> Integer.toString(location.getPartition()))
         .collect(Collectors.joining(","));
     // Add skip forward flag to properties
-    final Map<String, Object> requestProperties = ImmutableMap.of(
-        KsqlRequestConfig.KSQL_REQUEST_QUERY_PULL_SKIP_FORWARDING, true,
-        KsqlRequestConfig.KSQL_REQUEST_INTERNAL_REQUEST, true,
-        KsqlRequestConfig.KSQL_REQUEST_QUERY_PULL_PARTITIONS, partitions);
+    final Map<String, Object> requestProperties = new HashMap<>();
```
Comment (vpapavas) [inline ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java:339]: You can leave this as an `ImmutableMap` and use the `builder` to add an extra entry

Comment (vvcephei) [reply ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java:344]: +1

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java [hunk 2]
```diff
@@ -340,10 +341,14 @@ private static void forwardTo(
...
+    requestProperties.put(KsqlRequestConfig.KSQL_REQUEST_QUERY_PULL_SKIP_FORWARDING, true);
+    requestProperties.put(KsqlRequestConfig.KSQL_REQUEST_INTERNAL_REQUEST, true);
+    requestProperties.put(KsqlRequestConfig.KSQL_REQUEST_QUERY_PULL_PARTITIONS, partitions);
+    if (consistencyOffsetVector.isPresent()) {
+      requestProperties.put(KsqlRequestConfig.KSQL_REQUEST_QUERY_PULL_CONSISTENCY_OFFSET_VECTOR,
```
Comment (vvcephei) [inline ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java:349]: I think we also have to add the "consistency vector enabled" config, right, @vpapavas ?

Which raises the question: is this logic already being tested by the existing tests from your prior PR? (If so, and we are missing the other config, then how is that test passing?)

**Review by vvcephei — COMMENTED** [review]
Thanks, @patrickstuedi !

Just one thing to add on top of @vpapavas 's comment.

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java (hunk 2 above)
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java:349]: On the server side we check that:

1. The server config KsqlConfig.KSQL_QUERY_PULL_CONSISTENCY_OFFSET_VECTOR_ENABLED is set to true
2. The request properties contain a consistency token (may be an empty string the first time a request is made)

In the tests, @patrickstuedi has set the server property to true hence only the offset needs to be added.

We don't additionally need to check if the client has set the "consistency vector enabled" config, since the only way to add the vector to the request properties, is if this config is set. In this case here, the Optional will be non-empty, only if consistency is enabled

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+/**
+ * Test to ensure pull queries route across multiple KSQL nodes correctly.
+ *
+ * <p>For tests on general syntax and handled see RestQueryTranslationTest's
+ * materialized-aggregate-static-queries.json
+ */
+@SuppressWarnings("OptionalGetWithoutIsPresent")
+@Category({IntegrationTest.class})
+@RunWith(MockitoJUnitRunner.class)
+public class PullQueryConsistencyFunctionalTest {
```
Comment (vpapavas) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:98]: General comment about this class: There is a bunch of code that is copied from `PullQueryRoutingFunctionalTest`. How about moving the common code to the `HighAvailabilityTestUtil` class?

Comment (vvcephei) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:98]: This is objectively good advice.

Just FYI, I could also get behind keeping them independent so that the tests can evolve independently without affecting each other. I think the real deciding factor isn't how much code is similar between the two classes, but how much hassle do you anticipate in the future if we keep them independent or if we create a common util class?

For example, if there's a lot of common code, but it's all pretty straightforward, and you think it'll just sit there not hurting anyone for the next 5 years, then I'd say don't bother combining them. But if it looks like the common code is pretty tricky and you think any change to one test would imply we should also change the other test, then it's better to combine the common code.

In other words, you've got my opinion, and I'm happy whichever way you decide to go.

**Review by vvcephei — APPROVED** [review]
LGTM, mod @vpapavas 's open question.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java (hunk 3 above)
Comment (patrickstuedi) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:98]: Ok, thanks for both comments @vpapavas @vvcephei . Btw, one reason I didn't extend PullQueryRoutingFunctionalTest in the first place is because all those tests expect 1 partition, and I needed multiple partitions and there is no way to mock the "before" part depending on which tests is executed.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+      clusterFormation.router.getApp(),
+      clusterFormation.standBy.getHost(),
+      HighAvailabilityTestUtil::remoteServerIsUp, USER_CREDS);
+
+    // When:
+    final List<StreamedRow> rows_0 = makeRequestIncludingConsistencyToken(clusterFormation.router.getApp(),
+      sqlTableScan, null, USER_CREDS);
+
+    // Then:
+    assertThat(rows_0, hasSize(HEADER + 5));
```
Comment (guozhangwang) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:319]: Just curious, could we check the actual row content, more than just the size?

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+    public KsqlHostInfoEntity getHost() {
+      return host;
+    }
+
+    public TestKsqlRestApp getApp() {
+      return app;
+    }
+
+    public Shutoffs getShutoffs() {
```
Comment (guozhangwang) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:430]: This seems not used?

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java (hunk 3 above)
Comment (patrickstuedi) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:98]: @vpapavas I was playing around with moving common parts around, but it's tricky, you move one thing then we need to move another thing et etc, eventually we need to move TestApps and the actual number of it, etc, all things that we might want to change in the future and certain things are all already changed

I think it would be best to keep the copy of those infra calls currently

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java (hunk 5 above)
Comment (patrickstuedi) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:452]: Good catch @guozhangwang . The thing is, this test class is where lots of the consistency tests will eventually go in. Currently is just checking that forwarding and returning of the token works, but eventually we want to test consistency tokens are handled as expected, also in the presence of shutdown/rebalance/etc, so I prefer to keep those infra calls for the moment.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+      .skip(HEADER)
+      .filter(sr -> sr.getRow().isPresent())
+      .map(sr -> {
+        return sr.getRow().get().getColumns();
+      })
+      .collect(Collectors.toList());
+
+    // Then:
+    assertThat(rows, hasSize(HEADER + 6));
+    assertThat(rows.get(HEADER+5).getConsistencyToken(), is(not(Optional.empty())));
```
Comment (vpapavas) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java:331]: What you actually want to check is that the consistency token contains the information you expect i.e., it did not get garbled after all the fan out and response.

**Review by vpapavas — APPROVED** [review]
Thank you @patrickstuedi ! LGTM just add the check in the test that the CV is actually correct
